### PR TITLE
Dont propagate flags between different compilers

### LINKED ISSUE #2786: Changing compiler in dependency does not change compiler flags
So here's the situation. I have both GCC and PGI in my `compilers.yaml`. GCC does not have any compiler flags, but PGI does:
```yaml
compilers:
- compiler:
    environment: {}
    extra_rpaths: []
    flags: {}
    modules: []
    operating_system: centos6
    paths:
      cc:  /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gcc
      cxx: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/g++
      f77: /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gfortran
      fc:  /soft/spack/opt/spack/linux-x86_64/gcc-5.3.0/gcc-6.1.0-q2zosj3igepi3pjnqt74bwazmptr5gpj/bin/gfortran
    spec: gcc@6.1.0
    target: x86_64
- compiler:
    environment: {}
    extra_rpaths: []
    flags:
      cflags:   -tp=haswell,sandybridge
      cxxflags: -tp=haswell,sandybridge
      fflags:   -tp=haswell,sandybridge
    modules: []
    operating_system: centos6
    paths:
      cc:  /soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-4.4.7/pgi-16.5-pa5hji6khmosol3h7u65zhw6cy2dnekn/linux86-64/16.5/bin/pgcc
      cxx: /soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-4.4.7/pgi-16.5-pa5hji6khmosol3h7u65zhw6cy2dnekn/linux86-64/16.5/bin/pgc++
      f77: /soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-4.4.7/pgi-16.5-pa5hji6khmosol3h7u65zhw6cy2dnekn/linux86-64/16.5/bin/pgfortran
      fc:  /soft/spack-0.9.1/opt/spack/linux-centos6-x86_64/gcc-4.4.7/pgi-16.5-pa5hji6khmosol3h7u65zhw6cy2dnekn/linux86-64/16.5/bin/pgfortran
    spec: pgi@16.5-0
    target: x86_64
```
Now I want to build `openmpi` with PGI, but `libpciaccess` can't be built with PGI, so I tell Spack to build it with GCC instead:
```
$ spack spec openmpi %pgi ^libpciaccess%gcc
Input spec
--------------------------------
openmpi%pgi
    ^libpciaccess%gcc

Normalized
--------------------------------
openmpi%pgi
    ^hwloc
        ^libpciaccess%gcc
            ^libtool
                ^m4@1.4.6:
            ^pkg-config@0.9.0:
            ^util-macros

Concretized
--------------------------------
openmpi@2.0.1%pgi@16.5-0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge"+java~mxm~pmi+psm~psm2~slurm~sqlite3+thread_multiple+tm~verbs+vt arch=linux-centos6-x86_64
    ^hwloc@1.11.4%pgi@16.5-0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
        ^libpciaccess@0.13.4%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
            ^libtool@2.4.6%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
                ^m4@1.4.17%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge"+sigsegv arch=linux-centos6-x86_64
                    ^libsigsegv@2.10%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
            ^pkg-config@0.29.1%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge"+internal_glib arch=linux-centos6-x86_64
            ^util-macros@1.19.0%gcc@6.1.0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
    ^jdk@8u92-linux-x64%pgi@16.5-0 cflags="-tp=haswell,sandybridge" cxxflags="-tp=haswell,sandybridge" fflags="-tp=haswell,sandybridge" arch=linux-centos6-x86_64
```
The problem is that although Spack will now correctly build libpciaccess and its dependencies with GCC, it is still trying to use the PGI compiler flags, which will fail.

### PR DESCRIPTION
**Pull request by scheibelp**
Fixes #2786

Previously when a spec had no parents with an equivalent compiler, Spack would default to adding the compiler flags associated with the root of the DAG. This eliminates that default.

This also fixes an infinite recursion issue when a child spec has compiler flags that the root spec does not. This was triggered by the following in concretize.py:

```
if (flag in spec.root.compiler_flags and
    ((flag not in spec.compiler_flags) or
     sorted(spec.compiler_flags[flag]) !=
     sorted(spec.root.compiler_flags[flag]))):
```

If ```spec.compiler_flags[flag]``` had a flag that the root did not, this continually evaluated as true without changing the root.

### CODE REVIEW

File: lib/spack/spack/concretize.py [hunk 1]
```diff
@@ -377,61 +377,40 @@ def concretize_compiler_flags(self, spec):
...
-                    spec.compiler_flags[flag] = list(
-                        set(spec.compiler_flags[flag]) |
-                        set(compiler.flags[flag]))
+            config_flags = set(compiler.flags.get(flag, []))
+            flags = set(spec.compiler_flags.get(flag, []))
+            if (config_flags - flags):
+                spec.compiler_flags[flag] = list(config_flags | flags)
+                ret = True
+            elif flag not in spec.compiler_flags:
+                spec.compiler_flags[flag] = list()
```
Comment (becker33) [inline lib/spack/spack/concretize.py]: Shouldn't this be `spec.compiler_flags[flag] = config_flags`? I think this currently drops flags from the config file if that flag is not already in the spec.

Comment (scheibelp) [inline lib/spack/spack/concretize.py]: To get here, ```config_flags - flags``` must be an empty set and ```flags``` must be an empty set, so ```config_flags``` would also be an empty set

Comment (scheibelp) [inline lib/spack/spack/concretize.py]: Would it be better if I replaced:

```elif flag not in spec.compiler_flags```

with

```elif not flags```

?

Comment (becker33) [inline lib/spack/spack/concretize.py]: Yes, I know see that it works, it's just a clarity issue. What about replacing the whole `if` block with
```
spec.compiler_flags[flag] = list(config_flags | flags)
if (config_flags - flags):
    ret = True
````
That's "Add in everything from config_flags, and if we changed it set changed to True"

Comment (scheibelp) [inline lib/spack/spack/concretize.py]: Looks good! I'll get that updated in the next hour.

**Pull request merged**